### ISSUE TITLE
Cannot parse URI without authority component

### ISSUE THREAD

**Issue by maxlambrecht** [hyperium/http#469]
Trying to parse an URI like `unix:///tmp/agent.sock` or `unix:/tmp/agent.sock`, that has a schema and not an authority, produces an Error `InvalidaUri(InvalidFormat)`.

There is no restriction defined in the [URI standard](https://tools.ietf.org/html/rfc3986#section-3) that would make these URIs invalid.

**Comment by adamreichold** [comment]
I think I am hitting a similar issue where a location header that contains a path is parsed as an authority
```
[tests/path.rs:5] "forum".parse::<Uri>().unwrap().into_parts() = Parts {
    scheme: None,
    authority: Some(
        forum,
    ),
    path_and_query: None,
    _priv: (),
}
```
or not at all if it has a trailing slash
```
[tests/path.rs:5] "forum/".parse::<Uri>() = Err(
    InvalidUri(
        InvalidFormat,
    ),
)
```

**Comment by nox** [comment]
@adamreichold What you pasted is orthogonal to the ticket and AFAIK both your pastes aren't bugs.

**Comment by adamreichold** [comment]
@nox Thank you for looking into this! Can you go into slightly more detail why both instances are valid behaviour so that I can better understand how to work around this? Thanks!

**Comment by nox** [comment]
The `Uri` type in that crate was mostly designed to represent a HTTP request target as defined in https://httpwg.org/specs/rfc7230.html#request-target.

* `origin-form` is more or less an absolute path, that starts with `/`.
* `absolute-form` is a complete URL in all its glory, with the scheme etc.
* `authority-form` is well, an authority, that is, the `localhost` part in `http://localhost/`, that's your first example.
* `asterisk-form` is just `*`, used for `OPTIONS` requests for example.

So this covers the first example `forum`, which as you've stated is successfully.

If you read the rules for those 4 cases, you'll notice that the only forms that can accept a `/` are `origin-form` and `absolute-form`, and that `forum/` fits neither, hence the error you've encountered.

**Comment by adamreichold** [comment]
> The Uri type in that crate was mostly designed to represent a HTTP request target 

So basically, this boils down to `Uri` handling `request-target` but not `URI-reference` in the parlance of RFC 7230. :thinking:

**Comment by nox** [comment]
Yes, cause it represents an URI in the context of a HTTP request/response cycle, where `GET forum HTTP/1.0` is not valid.

**Comment by 89z** [comment]
https://github.com/adamreichold/zeptohttpc/issues/6

**Comment by 89z** [comment]
@nox are you sure about your position? A quick check of both Go and JavaScript, shows that both handle `forum/` just fine

**Comment by 89z** [comment]
Reading from RFC 3986 [1], Path is defined as:

~~~
path    = path-abempty    ; begins with "/" or is empty
           / path-absolute   ; begins with "/" but not "//"
           / path-noscheme   ; begins with a non-colon segment
           / path-rootless   ; begins with a segment
           / path-empty      ; zero characters
~~~

where `forum/` would apply to:

~~~
path-noscheme = segment-nz-nc *( "/" segment )
path-rootless = segment-nz *( "/" segment )
~~~

lets just focus on the second one:

~~~
path-rootless = segment-nz *( "/" segment )
~~~

more definitions:

~~~
segment       = *pchar
segment-nz    = 1*pchar
~~~

more definitions [2]:

~~~
*pchar = allows any number, including zero
1*pchar = requires at least one
~~~

So, according to that, these would all be valid:

~~~
hello
hello/
hello/world
~~~

Finally, the canonical definition of URIs is in RFC 3986, not RFC 7230. This is evidenced by the fact that RFC 3986 is mentioned at least 30 times in RFC 7230, notably in the section Uniform Resource Identifiers [3]. In conclusion, I think @nox should reconsider their position.

1. https://rfc-editor.org/rfc/rfc3986#section-3.3
2. https://rfc-editor.org/rfc/rfc2616#section-2.1
3. https://rfc-editor.org/rfc/rfc7230.html#section-2.7
